Wrap "db:seed" task in transaction

davidpdrsn opened the pull request:
Fixes #33302

If running your seeds file failed halfway through for some reason you could be
left with a database in an inconsistent state. That is fixed by wrapping it in
a transaction.

I guess it might be nice to tell the user that the changes were rolled back.
Let me know if I should add that 😄

davidpdrsn committed:
Wrap "db:seed" task in transaction

Fixes #33302

If running your seeds file failed for some reason you could be left with
a database in an inconsistent state. That is fixed by wrapping it in a
transaction.

twitnithegirl commented:
@davidpdrsn as an end-user I think you're right that it would be better to have a message for the failure/rollback. The behavior should also be tested somewhere. Looks like database tasks are being tested here https://github.com/rails/rails/blob/master/activerecord/test/cases/tasks/database_tasks_test.rb
Though, the `rake db:seed` task isn't being tested anywhere at the moment 🤷

davidpdrsn committed:
Tell the user that the seeds transaction got rolled back

By doing this users will know for sure that they didn't leave the
database in an inconsistent state.

I moved the opening of the transaction from
"activerecord/lib/active_record/railties/databases.rake" to
"railties/lib/rails/engine.rb" because there already were tests for the seeds
file being loaded there, that I could base my tests on.

I suppose modifying `$stderr` is generally a bad idea, but it was the only way
I could think off to also verify that some output was given to the user. I
consider this output to be important and worth testing, so I decided it was a
worthy trade-off.

davidpdrsn commented:
I have now added some output to the user when something goes wrong and added tests. See [the commit message](https://github.com/rails/rails/pull/33356/commits/6dd2b9c3bcb2bb4064b8df67ec5ab9cf4d3dd11b) for more details.

kaspth reviewed (commented):
Thanks! Makes fine sense to me to wrap the loading in a transaction.

kamipo commented on railties/lib/rails/engine.rb:
```diff
@@ -548,7 +548,18 @@ def config
+          raise
+        end
+      end
     end
```
`ActiveRecord::Base.transaction` would basically roll back only the changes on the "primary" connection spec.
So "All changes rolled back." may makes a misunderstanding to users.

kaspth replied:
Good point. This should protect a database regardless of whether it's primary or secondary. Aren't we connected to the secondary spec here if we're seed loading on a secondary db though?

davidpdrsn replied:
Not sure what you mean by primary or secondary connection 🤔 

Regarding moving the code to the database tasks, I tried doing that at first but was unsure about how to test it. "activerecord/test/cases/tasks/database_tasks_test.rb" doesn't include any seed tests that I could copy from. Any tips on how to test it?

twitnithegirl replied:
https://github.com/rails/rails/blob/master/activerecord/test/cases/tasks/database_tasks_test.rb has tests for other tasks so you could modify one of those to make a test for the seed tasks. I'd do a test for if the seed works and a test for if it fails.

kaspth replied:
We usually don't let testing determine where our lib code should live. Let's see if we can test it in database_tasks_test.rb as you and @twitnithegirl mentioned 😄

davidpdrsn replied:
Right I agree. I'll try to work on it tomorrow.

davidpdrsn commented on railties/test/railties/engine_test.rb:
```diff
@@ -879,6 +879,56 @@ def index
+      app_file "db/seeds.rb", <<-RUBY
+        unless ActiveRecord::Base.connection.transaction_open?
+          raise "Not in transaction"
+        end
```
Yes. The test needs to somehow fail if it doesn't wrap the seed in a transaction. This was the only I could think of testing that 😄

davidpdrsn committed:
Remove needless `rescue`

davidpdrsn committed:
Use `assert_includes`

davidpdrsn committed:
Refactor error message assertions

davidpdrsn committed:
Wrap overriding `$stderr` to avoid strange behavior down the road

rafaelfranca commented on railties/lib/rails/engine.rb:
```diff
@@ -548,7 +548,18 @@ def config
       seed_file = paths["db/seeds.rb"].existent.first
-      load(seed_file) if seed_file
+      return unless seed_file
+      ActiveRecord::Base.transaction do
```
railties don't depend on Active Record so we can't call methods of Active Record here. I think a different way to fix this problem is to generate the transaction in the seed file. This way the user chan chose which database connection they want to use and we don't make railtie depend on Active Record. Or move to database tasks as recommended by Kasper

kaspth commented:
I think this is obviated by `bin/rails db:seed:replant`. Now, in case seeding fails, it's easy to clear the database and try again.

Wrapping the whole seeding in a single transaction is also tough for multiple dbs. That prevents opening a connection to another database in db/seeds and putting records in there.

Thanks for the PR!
